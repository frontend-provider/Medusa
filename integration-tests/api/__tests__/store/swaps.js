const path = require("path")
const {
  ShippingProfile,
  ShippingOption,
  ShippingProfileType,
} = require("@medusajs/medusa")

const setupServer = require("../../../helpers/setup-server")
const { useApi } = require("../../../helpers/use-api")
const { initDb, useDb } = require("../../../helpers/use-db")

const orderSeeder = require("../../helpers/order-seeder")

jest.setTimeout(30000)

describe("/store/carts", () => {
  let medusaProcess
  let dbConnection

  beforeAll(async () => {
    const cwd = path.resolve(path.join(__dirname, "..", ".."))
    dbConnection = await initDb({ cwd })
    medusaProcess = await setupServer({ cwd })
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()
    medusaProcess.kill()
  })

  describe("/store/swaps", () => {
    beforeEach(async () => {
      await orderSeeder(dbConnection)

      const manager = dbConnection.manager
      await manager.query(
        `UPDATE "swap" SET cart_id='test-cart-2' WHERE id = 'test-swap'`
      )
      await manager.query(
        `UPDATE "payment" SET swap_id=NULL WHERE id = 'test-payment-swap'`
      )

      const defaultProfile = await manager.findOne(ShippingProfile, {
        where: {
          type: ShippingProfileType.DEFAULT,
        },
      })
      await manager.insert(ShippingOption, {
        id: "return-option",
        name: "Test ret",
        profile_id: defaultProfile.id,
        region_id: "test-region",
        provider_id: "test-ful",
        data: {},
        price_type: "flat_rate",
        amount: 1000,
        is_return: true,
      })
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("simple swap", async () => {
      const api = useApi()

      const response = await api
        .post("/store/swaps", {
          order_id: "test-order",
          return_items: [
            {
              item_id: "test-item",
              quantity: 1,
            },
          ],
          additional_items: [
            {
              variant_id: "test-variant-2",
              quantity: 1,
            },
          ],
        })
        .catch((err) => {
          console.log(err.response.data.message)
        })

      expect(response.data).toEqual(
        expect.objectContaining({
          swap: expect.objectContaining({
            id: expect.stringMatching(/^swap_*/),
            additional_items: expect.arrayContaining([
              expect.objectContaining({
                id: expect.stringMatching(/^item_*/),
                cart_id: expect.stringMatching(/^cart_*/),
                swap_id: expect.stringMatching(/^swap_*/),
                variant: expect.objectContaining({
                  id: "test-variant-2",
                }),
                quantity: 1,
                variant_id: "test-variant-2",
              }),
            ]),
            order: expect.objectContaining({
              id: "test-order",
            }),
            cart_id: expect.stringMatching(/^cart_*/),
            cart: expect.objectContaining({
              id: expect.stringMatching(/^cart_*/),
              billing_address_id: "test-billing-address",
              type: "swap",
              shipping_address_id: "test-shipping-address",
              metadata: expect.objectContaining({
                swap_id: expect.stringMatching(/^swap_*/),
              }),
            }),
            return_order: expect.objectContaining({
              id: expect.stringMatching(/^ret_*/),
              swap_id: expect.stringMatching(/^swap_*/),
              refund_amount: 7200,
              items: expect.arrayContaining([
                expect.objectContaining({
                  item_id: "test-item",
                  quantity: 1,
                  return_id: expect.stringMatching(/^ret_*/),
                }),
              ]),
            }),
          }),
        })
      )
    })

    it("swap with return shipping", async () => {
      const api = useApi()

      const response = await api
        .post("/store/swaps", {
          order_id: "test-order",
          return_items: [
            {
              item_id: "test-item",
              quantity: 1,
            },
          ],
          return_shipping_option: "return-option",
          additional_items: [
            {
              variant_id: "test-variant-2",
              quantity: 1,
            },
          ],
        })
        .catch((err) => {
          console.log(err.response.data.message)
        })

      expect(response.data).toEqual({
        swap: expect.objectContaining({
          id: expect.stringMatching(/^swap_*/),
          additional_items: expect.arrayContaining([
            expect.objectContaining({
              id: expect.stringMatching(/^item_*/),
              cart_id: expect.stringMatching(/^cart_*/),
              swap_id: expect.stringMatching(/^swap_*/),
              variant: expect.objectContaining({
                id: "test-variant-2",
              }),
              quantity: 1,
              variant_id: "test-variant-2",
            }),
          ]),
          order: expect.objectContaining({
            id: "test-order",
          }),
          cart_id: expect.stringMatching(/^cart_*/),
          cart: expect.objectContaining({
            id: expect.stringMatching(/^cart_*/),
            billing_address_id: "test-billing-address",
            shipping_address_id: "test-shipping-address",
            type: "swap",
            metadata: expect.objectContaining({
              swap_id: expect.stringMatching(/^swap_*/),
            }),
          }),
          return_order: expect.objectContaining({
            id: expect.stringMatching(/^ret_*/),
            swap_id: expect.stringMatching(/^swap_*/),
            refund_amount: 6200,
            shipping_method: expect.objectContaining({
              id: expect.stringMatching(/^sm_*/),
              return_id: expect.stringMatching(/^ret_*/),
              shipping_option: expect.objectContaining({
                profile_id: expect.stringMatching(/^sp_*/),
              }),
            }),
            items: expect.arrayContaining([
              expect.objectContaining({
                item_id: "test-item",
                quantity: 1,
                return_id: expect.stringMatching(/^ret_*/),
              }),
            ]),
          }),
        }),
      })
    })
  })
})
