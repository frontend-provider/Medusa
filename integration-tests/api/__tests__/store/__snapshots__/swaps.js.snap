// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/store/carts /store/swaps simple swap 1`] = `
Object {
  "swap": Object {
    "additional_items": Array [
      Object {
        "allow_discounts": true,
        "cart_id": StringMatching /\\^cart_\\*/,
        "claim_order_id": null,
        "created_at": Any<String>,
        "description": "Swap product",
        "fulfilled_quantity": null,
        "has_shipping": null,
        "id": StringMatching /\\^item_\\*/,
        "is_giftcard": false,
        "is_return": false,
        "metadata": Object {},
        "order_id": null,
        "quantity": 1,
        "returned_quantity": null,
        "shipped_quantity": null,
        "should_merge": true,
        "swap_id": StringMatching /\\^swap_\\*/,
        "thumbnail": null,
        "title": "test product",
        "unit_price": 8000,
        "updated_at": Any<String>,
        "variant": Object {
          "allow_backorder": false,
          "barcode": null,
          "created_at": Any<String>,
          "deleted_at": null,
          "ean": null,
          "height": null,
          "hs_code": null,
          "id": "test-variant-2",
          "inventory_quantity": 1,
          "length": null,
          "manage_inventory": true,
          "material": null,
          "metadata": null,
          "mid_code": null,
          "origin_country": null,
          "product": Object {
            "collection_id": null,
            "created_at": Any<String>,
            "deleted_at": null,
            "description": null,
            "discountable": true,
            "external_id": null,
            "handle": null,
            "height": null,
            "hs_code": null,
            "id": "test-product",
            "is_giftcard": false,
            "length": null,
            "material": null,
            "metadata": null,
            "mid_code": null,
            "origin_country": null,
            "profile_id": StringMatching /\\^sp_\\*/,
            "status": "draft",
            "subtitle": null,
            "thumbnail": null,
            "title": "test product",
            "type_id": null,
            "updated_at": Any<String>,
            "weight": null,
            "width": null,
          },
          "product_id": "test-product",
          "sku": null,
          "title": "Swap product",
          "upc": null,
          "updated_at": Any<String>,
          "weight": null,
          "width": null,
        },
        "variant_id": "test-variant-2",
      },
    ],
    "cart": Object {
      "billing_address_id": "test-billing-address",
      "completed_at": null,
      "context": Object {},
      "created_at": Any<String>,
      "customer_id": "test-customer",
      "deleted_at": null,
      "email": "test@email.com",
      "id": StringMatching /\\^cart_\\*/,
      "idempotency_key": null,
      "metadata": Object {
        "parent_order_id": "test-order",
        "swap_id": StringMatching /\\^swap_\\*/,
      },
      "object": "cart",
      "payment_authorized_at": null,
      "payment_id": null,
      "region_id": "test-region",
      "shipping_address_id": "test-shipping-address",
      "type": "swap",
      "updated_at": Any<String>,
    },
    "cart_id": StringMatching /\\^cart_\\*/,
    "confirmed_at": null,
    "created_at": Any<String>,
    "deleted_at": null,
    "difference_due": null,
    "fulfillment_status": "not_fulfilled",
    "fulfillments": Array [],
    "id": StringMatching /\\^swap_\\*/,
    "idempotency_key": Any<String>,
    "metadata": null,
    "order": Object {
      "billing_address_id": "test-billing-address",
      "canceled_at": null,
      "cart_id": null,
      "created_at": Any<String>,
      "currency_code": "usd",
      "customer_id": "test-customer",
      "display_id": 1,
      "draft_order_id": null,
      "email": "test@email.com",
      "external_id": null,
      "fulfillment_status": "fulfilled",
      "id": "test-order",
      "idempotency_key": null,
      "metadata": null,
      "no_notification": null,
      "object": "order",
      "payment_status": "captured",
      "region_id": "test-region",
      "shipping_address_id": "test-shipping-address",
      "status": "pending",
      "tax_rate": 0,
      "updated_at": Any<String>,
    },
    "order_id": "test-order",
    "payment": null,
    "payment_status": "not_paid",
    "return_order": Object {
      "claim_order_id": null,
      "created_at": Any<String>,
      "id": StringMatching /\\^ret_\\*/,
      "idempotency_key": null,
      "items": Array [
        Object {
          "is_requested": true,
          "item_id": "test-item",
          "metadata": null,
          "note": null,
          "quantity": 1,
          "reason_id": null,
          "received_quantity": null,
          "requested_quantity": 1,
          "return_id": StringMatching /\\^ret_\\*/,
        },
      ],
      "metadata": null,
      "no_notification": true,
      "order_id": null,
      "received_at": null,
      "refund_amount": 7200,
      "shipping_data": null,
      "shipping_method": null,
      "status": "requested",
      "swap_id": StringMatching /\\^swap_\\*/,
      "updated_at": Any<String>,
    },
    "shipping_address": null,
    "shipping_address_id": null,
    "shipping_methods": Array [],
    "updated_at": Any<String>,
  },
}
`;

exports[`/store/carts /store/swaps swap with return shipping 1`] = `
Object {
  "swap": Object {
    "additional_items": Array [
      Object {
        "allow_discounts": true,
        "cart_id": StringMatching /\\^cart_\\*/,
        "claim_order_id": null,
        "created_at": Any<String>,
        "description": "Swap product",
        "fulfilled_quantity": null,
        "has_shipping": null,
        "id": StringMatching /\\^item_\\*/,
        "is_giftcard": false,
        "is_return": false,
        "metadata": Object {},
        "order_id": null,
        "quantity": 1,
        "returned_quantity": null,
        "shipped_quantity": null,
        "should_merge": true,
        "swap_id": StringMatching /\\^swap_\\*/,
        "thumbnail": null,
        "title": "test product",
        "unit_price": 8000,
        "updated_at": Any<String>,
        "variant": Object {
          "allow_backorder": false,
          "barcode": null,
          "created_at": Any<String>,
          "deleted_at": null,
          "ean": null,
          "height": null,
          "hs_code": null,
          "id": "test-variant-2",
          "inventory_quantity": 1,
          "length": null,
          "manage_inventory": true,
          "material": null,
          "metadata": null,
          "mid_code": null,
          "origin_country": null,
          "product": Object {
            "collection_id": null,
            "created_at": Any<String>,
            "deleted_at": null,
            "description": null,
            "discountable": true,
            "external_id": null,
            "handle": null,
            "height": null,
            "hs_code": null,
            "id": "test-product",
            "is_giftcard": false,
            "length": null,
            "material": null,
            "metadata": null,
            "mid_code": null,
            "origin_country": null,
            "profile_id": StringMatching /\\^sp_\\*/,
            "status": "draft",
            "subtitle": null,
            "thumbnail": null,
            "title": "test product",
            "type_id": null,
            "updated_at": Any<String>,
            "weight": null,
            "width": null,
          },
          "product_id": "test-product",
          "sku": null,
          "title": "Swap product",
          "upc": null,
          "updated_at": Any<String>,
          "weight": null,
          "width": null,
        },
        "variant_id": "test-variant-2",
      },
    ],
    "cart": Object {
      "billing_address_id": "test-billing-address",
      "completed_at": null,
      "context": Object {},
      "created_at": Any<String>,
      "customer_id": "test-customer",
      "deleted_at": null,
      "email": "test@email.com",
      "id": StringMatching /\\^cart_\\*/,
      "idempotency_key": null,
      "metadata": Object {
        "parent_order_id": "test-order",
        "swap_id": StringMatching /\\^swap_\\*/,
      },
      "object": "cart",
      "payment_authorized_at": null,
      "payment_id": null,
      "region_id": "test-region",
      "shipping_address_id": "test-shipping-address",
      "type": "swap",
      "updated_at": Any<String>,
    },
    "cart_id": StringMatching /\\^cart_\\*/,
    "confirmed_at": null,
    "created_at": Any<String>,
    "deleted_at": null,
    "difference_due": null,
    "fulfillment_status": "not_fulfilled",
    "fulfillments": Array [],
    "id": StringMatching /\\^swap_\\*/,
    "idempotency_key": Any<String>,
    "metadata": null,
    "order": Object {
      "billing_address_id": "test-billing-address",
      "canceled_at": null,
      "cart_id": null,
      "created_at": Any<String>,
      "currency_code": "usd",
      "customer_id": "test-customer",
      "display_id": Any<Number>,
      "draft_order_id": null,
      "email": "test@email.com",
      "external_id": null,
      "fulfillment_status": "fulfilled",
      "id": "test-order",
      "idempotency_key": null,
      "metadata": null,
      "no_notification": null,
      "object": "order",
      "payment_status": "captured",
      "region_id": "test-region",
      "shipping_address_id": "test-shipping-address",
      "status": "pending",
      "tax_rate": 0,
      "updated_at": Any<String>,
    },
    "order_id": "test-order",
    "payment": null,
    "payment_status": "not_paid",
    "return_order": Object {
      "claim_order_id": null,
      "created_at": Any<String>,
      "id": StringMatching /\\^ret_\\*/,
      "idempotency_key": null,
      "items": Array [
        Object {
          "is_requested": true,
          "item_id": "test-item",
          "metadata": null,
          "note": null,
          "quantity": 1,
          "reason_id": null,
          "received_quantity": null,
          "requested_quantity": 1,
          "return_id": StringMatching /\\^ret_\\*/,
        },
      ],
      "metadata": null,
      "no_notification": true,
      "order_id": null,
      "received_at": null,
      "refund_amount": 6200,
      "shipping_data": Object {},
      "shipping_method": Object {
        "cart_id": null,
        "claim_order_id": null,
        "data": Object {},
        "id": StringMatching /\\^sm_\\*/,
        "order_id": null,
        "price": 1000,
        "return_id": StringMatching /\\^ret_\\*/,
        "shipping_option": Object {
          "admin_only": false,
          "amount": 1000,
          "created_at": Any<String>,
          "data": Object {},
          "deleted_at": null,
          "id": "return-option",
          "is_return": true,
          "metadata": null,
          "name": "Test ret",
          "price_type": "flat_rate",
          "profile_id": StringMatching /\\^sp_\\*/,
          "provider_id": "test-ful",
          "region_id": "test-region",
          "updated_at": Any<String>,
        },
        "shipping_option_id": "return-option",
        "swap_id": null,
      },
      "status": "requested",
      "swap_id": StringMatching /\\^swap_\\*/,
      "updated_at": Any<String>,
    },
    "shipping_address": null,
    "shipping_address_id": null,
    "shipping_methods": Array [],
    "updated_at": Any<String>,
  },
}
`;
